src/main/java/teammates/logic/api/AccessControlUtil.java

```diff
@@ -1,5 +1,214 @@
+        }
+        return permissions;
+    }
+    
+    private boolean isStudentRecipientOfComment(CommentAttributes comment, StudentAttributes student) {
+        return CommentParticipantType.PERSON.equals(comment.recipientType) && comment.recipients.contains(student.email);
+    }
+    
+    private boolean isStudentTeamRecipientOfComment(CommentAttributes comment, StudentAttributes student) {
+        return CommentParticipantType.TEAM.equals(comment.recipientType)
+                && comment.recipients.contains(Sanitizer.sanitizeForHtml(student.team));
+    }
+    
+    private boolean isStudentInSameTeamAsRecipientOfComment(
+            CommentAttributes comment, StudentAttributes student, List<String> teammatesEmails) {
+        // if recipient type is not a student, this comment is not directed at a teammate
+        if (!CommentParticipantType.PERSON.equals(comment.recipientType)) {
+            return false;
+        }
+        
+        return isCommentRecipientPresentInList(comment, teammatesEmails);
+    }
+    
+    private boolean isStudentSectionRecipientOfComment(CommentAttributes comment, StudentAttributes student) {
+        return CommentParticipantType.SECTION.equals(comment.recipientType) && comment.recipients.contains(student.section);
+    }
+    
+    private boolean isStudentInSameSectionAsRecipientOfComment(
+            CommentAttributes comment, StudentAttributes student,
+            List<String> sectionStudentsEmails, List<String> teamsInSection) {
+        // if recipient type is not a student or team, this comment is not directed at a recipient in the
+        // same section as the student
+        if (!CommentParticipantType.PERSON.equals(comment.recipientType)
+                && !CommentParticipantType.TEAM.equals(comment.recipientType)) {
+            return false;
+        }
+        
+        // check if student is in same section
+        if (CommentParticipantType.PERSON.equals(comment.recipientType)
+                && isCommentRecipientPresentInList(comment, sectionStudentsEmails)) {
+            return true;
+            
+        // check if team is in same section
+        } else if (isCommentRecipientPresentInList(comment, teamsInSection)) {
+            return true;
+        }
+        
+        return false;
+    }
+    
+    private boolean isStudentCourseRecipientOfComment(CommentAttributes comment, StudentAttributes student) {
+        return CommentParticipantType.COURSE.equals(comment.recipientType) && comment.courseId.equals(student.course);
+    }
+    
+    private boolean isStudentInSameCourseAsRecipientOfComment(CommentAttributes comment, StudentAttributes student) {
+        return comment.courseId.equals(student.course);
+    }
+    
+    private boolean isCommentRecipientPresentInList(CommentAttributes comment, List<String> candidateRecipients) {
+        if (!comment.recipients.isEmpty()) {
+            String commentRecipient = comment.recipients.iterator().next();
```

Comment:
Different from the original implementation in `CommentsLogic`, which assumed the possibility of multiple recipients. Can be easily changed if required.